tests/test_restrict_hosts.py

```diff
@@ -0,0 +1,235 @@
+# -*- coding: utf-8 -*-
+import pytest
+
+try:
+    from urllib.parse import urlparse
+except ImportError:
+    from urlparse import urlparse
+
+import inspect
+
+
+localhost = '127.0.0.1'
+
+connect_code_template = """
+    import socket
+    import pytest
+
+    {3}
+    def {2}():
+        socket.socket().connect(('{0}', {1}))
+"""
+
+urlopen_code_template = """
+    import pytest
+    try:
+        from urllib.request import urlopen
+    except ImportError:
+        from urllib2 import urlopen
+
+    {3}
+    def {2}():
+        assert urlopen('http://{0}:{1}/').getcode() == 200
+"""
+
+
+def assert_host_blocked(result, host):
+    result.stdout.fnmatch_lines('*A test tried to use socket.socket.connect() with host "{0}"*'.format(host))
+
+
+@pytest.fixture
+def assert_connect(testdir, httpserver):
+    def assert_socket_connect(should_pass, **kwargs):
```

Comment:
What is the benefit of nesting this method inside `assert_connect`? It feels like a confusing approach for a method that is only called once.

Reply:
It's just a way of returning a function to be executed in the test. By nesting it as a fixture i can have the other fixtures injected and return a function which only takes what is different per-test. For me it makes the tests very descriptive.